File: samples/java-valueentity-counter-code-first/src/main/java/com/example/domain/Counter.java
/* This code was initialised by Akka Serverless tooling.
 * As long as this file exists it will not be re-generated.
 * You are free to make changes to this file.
 */

package com.example.domain;

import kalix.javasdk.valueentity.ValueEntity;

public class Counter extends ValueEntity<CounterState> {

    @Override
    public CounterState emptyState() {
      return new CounterState(0);
    }
    public Effect<Response> increase(CounterState currentState, Increase command) {
      if (command.increaseBy < 0) {
        return effects().error("Increase requires a positive value. It was [" + command.increaseBy + "].");
      }
      return effects()
              .updateState(currentState.increase(command.increaseBy))
              .thenReply(Response.done());
    }

    public Effect<Response> decrease(CounterState currentState, Decrease command) {
      if (command.decreaseBy < 0) {
        return effects().error("Decrease requires a positive value. It was [" + command.decreaseBy + "].");
      }
      return effects()
          .updateState(currentState.decrease(command.decreaseBy))
          .thenReply(Response.done());
    }
    
    public Effect<Response> reset(CounterState currentState, Reset command) {
      return effects()
          .updateState(emptyState())
          .thenReply(Response.done());
    }
    
    public Effect<CurrentCounter> getCurrentCounter(CounterState currentState, GetCounter command) {
        CurrentCounter current = new CurrentCounter(currentState.value);
        return effects().reply(current);
    }
}

File: sbt-plugin/src/sbt-test/sbt-kalix/eventsourcedentity/src/main/scala/com/example/eventsourcedentity/domain/Counter.scala
package com.example.eventsourcedentity.domain

import kalix.scalasdk.eventsourcedentity.EventSourcedEntity
import kalix.scalasdk.eventsourcedentity.EventSourcedEntityContext
import com.example.eventsourcedentity
import com.google.protobuf.empty.Empty

// This class was initially generated based on the .proto definition by Kalix tooling.
//
// As long as this file exists it will not be overwritten: you can maintain it yourself,
// or delete it so it is regenerated as needed.

class Counter(context: EventSourcedEntityContext) extends AbstractCounter {
  override def emptyState: CounterState =
    CounterState(0)

  override def increase(currentState: CounterState, increaseValue: eventsourcedentity.IncreaseValue): EventSourcedEntity.Effect[Empty] =
    effects.emitEvent(Increased(increaseValue.value)).thenReply(_ => Empty())

  override def decrease(currentState: CounterState, decreaseValue: eventsourcedentity.DecreaseValue): EventSourcedEntity.Effect[Empty] =
    effects.emitEvent(Decreased(decreaseValue.value)).thenReply(_ => Empty())

  override def reset(currentState: CounterState, resetValue: eventsourcedentity.ResetValue): EventSourcedEntity.Effect[Empty] =
    if (currentState.value == 0)
      effects.reply(Empty())
    else if (currentState.value > 0)
      effects.emitEvent(Decreased(currentState.value)).thenReply(_ => Empty())
    else
      effects.emitEvent(Increased(-currentState.value)).thenReply(_ => Empty())

  override def getCurrentCounter(currentState: CounterState, getCounter: eventsourcedentity.GetCounter): EventSourcedEntity.Effect[eventsourcedentity.CurrentCounter] =
    effects.reply(eventsourcedentity.CurrentCounter(currentState.value))

  override def increased(currentState: CounterState, increased: Increased): CounterState =
    currentState.copy(value = currentState.value + increased.value)

  override def decreased(currentState: CounterState, decreased: Decreased): CounterState =
    currentState.copy(value = currentState.value - decreased.value)
}
